### ISSUE TITLE
[OSSHACK] Document Thumbnails

### ISSUE THREAD

**Issue by Mythie** [documenso/documenso#704]
## Improvement Description

Add a thumbnail generation feature for documents uploaded, thumbnails will be created using the first page of the uploaded document.

A high quality and low quality thumbnail should be generated for each document uploaded. The low quality thumbnail will be used in the documents table for quick viewing, while the high quality thumbnail will be used for a more detailed preview upon clicking the thumbnail.

## Rationale

Right now, the documents table is very plain and does not provide any visual information about the document. Thumbnails will allow users to quickly identify the document they are looking for.

This is extremely useful for users with many documents that have similar names since it provides a visual cue to help them find the document they are looking for.

## Proposed Solution

- Add a thumbnail generation feature for documents uploaded, thumbnails will be created using the first page of the uploaded document.
- Use either a background job or an endpoint to perform an upsert-like operation for retrieving the thumbnail.
- Store the thumbnails somewhere? Currently we have the univeral upload which is catered towards storing documents but could be used for storing thumbnails as well. Alternatively we also store data uris in the database which I'm less inclined to do but is an option.

## Alternatives

- Use a third party service to generate thumbnails. This would be the easiest solution but would require a third party service to be used which we would like to avoid.

## Additional Context

- While this might be a good candidate for background jobs we should consider how we can make this work without background jobs since we don't currently have a background job system in place.
- We would like to avoid third-party services since we would like to keep the project as self-contained as possible.

**Comment by itzabhinavarya** [comment]
@Mythie Assign this to me

**Comment by itzabhinavarya** [comment]
@Mythie  I am working on this.

**Comment by itzabhinavarya** [comment]
Hii @Mythie Sir, Will you share a loom video for reference.

**Comment by surajgjadhav** [comment]
HI @Mythie ,
can we use pdf.js for this feature?
or it is preferred to try fist without any external lib ?

**Comment by Mythie** [comment]
PDF.js is fine as we already use it in the app, although I'd like to suggest using https://unjs.io/packages/unpdf/ if you're heading in that direction since we will likely swap to it in the future.

**Comment by Mythie** [comment]
Do note that we will want to store the images rather than trying to render them using PDF.js during each page load since that can be detrimental to device performance.

If we are storing the images you'll need to consider how to use the canvas package on serverless environments for server side image generation or consider other flows 😄 

This is a bit open ended but just dropping thoughts while they're in my head!

**Comment by surajgjadhav** [comment]
Yes, from performance point of view we should limit only required data from the serverless 😅

**Comment by Mythie** [comment]
**Forwarding from Discord**

So to summarise, if I were doing a dirty implementation for a hackathon

- New table `DocumentThumbnail` which would have `id` `thumbnail_bytes` `hq_thumbnail_bytes`
- Store thumbnail on the `Document` via a foreign key
- Create a `/thumbnail/<id>` endpoint which would also support a `/thumbnail/<id>/hq`
- Use stale while revalidate to keep it cached for ages to reduce the performance impact

**Comment by ElTimuro** [comment]
removed oss hack tag
